File: Problemset/min-stack/min-stack.go

// @Title: 最小栈 (Min Stack)
// @Author: 846188037@qq.com
// @Date: 2020-10-20 11:20:03
// @Runtime: 16 ms
// @Memory: 10 MB

type MinStack struct {
    stack []int
    min []int

}


/** initialize your data structure here. */
func Constructor() MinStack {
    return MinStack{
        stack:make([]int,0),
        min:make([]int,0),
    }

}


func (this *MinStack) Push(x int)  {
    min:=this.GetMin()
    if x<=min{
        this.min=append(this.min,x)
    }
    this.stack=append(this.stack,x)

}


func (this *MinStack) Pop()  {
    if len(this.stack)==0{
        return 
    }
    if this.GetMin() == this.Top(){
        this.min=this.min[:len(this.min)-1]
    }
    this.stack=this.stack[:len(this.stack)-1]
}


func (this *MinStack) Top() int {
    if len(this.stack)==0{
        return 0
    }
    return this.stack[len(this.stack)-1]

}


func (this *MinStack) GetMin() int {
    if len(this.min)==0{
        return 1<<31
    }
    return this.min[len(this.min)-1]

}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: Problemset/min-stack/min-stack.java

// @Title: 最小栈 (Min Stack)
// @Author: 846188037@qq.com
// @Date: 2021-03-31 17:48:21
// @Runtime: 7 ms
// @Memory: 40.3 MB

class MinStack {

    /** initialize your data structure here. */
    private Stack<Integer> stackData;
    private Stack<Integer> stackMin;
    public MinStack() {
        this.stackData=new Stack<Integer>();
        this.stackMin=new Stack<Integer>();

    }
    
    public void push(int val) {
        if (this.stackMin.isEmpty()){
            this.stackMin.push(val);
        }else if(val<=this.getMin()){
            this.stackMin.push(val);
        }
        this.stackData.push(val);

    }
    
    public void pop() {
        if (this.stackData.isEmpty()){
            throw new RuntimeException("Your stack is a Empty");
        }
        int value=stackData.pop();
        if (value==this.getMin()){
            this.stackMin.pop();
        }
        //return value;

    }
    
    public int top() {
        if (this.stackData.isEmpty()){
            throw new RuntimeException("Your stack is empty");
        }
        return this.stackData.peek();

    }
    
    public int getMin() {
        if (this.stackMin.isEmpty()){
            throw new RuntimeException("Your Stack is empty");
        }
        return this.stackMin.peek();

    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
